src/dag-link.js

```diff
@@ -2,28 +2,57 @@
 
 const mh = require('multihashes')
 
+const ImmutableError = new Error('Immutable property')
+
 // Link represents an IPFS Merkle DAG Link between Nodes.
 class DAGLink {
   constructor (name, size, multihash) {
-    this.name = name
-    this.size = size
+    this._name = name
+    this._size = size
+    this._multihash = null
 
     if (typeof multihash === 'string') {
-      this.multihash = mh.fromB58String(multihash)
+      this._multihash = mh.fromB58String(multihash)
     } else if (Buffer.isBuffer(multihash)) {
-      this.multihash = multihash
+      this._multihash = multihash
     }
+  }
+
+  toString () {
+    const mhStr = mh.toB58String(this.multihash)
+    return `DAGLink <${mhStr} - name: "${this.name}", size: ${this.size}>`
+  }
 
-    this.json = {
+  toJSON () {
+    return {
       name: this.name,
       size: this.size,
-      hash: this.multihash ? mh.toB58String(this.multihash) : undefined
+      hash: this.multihash ? mh.toB58String(this.multihash) : null
     }
   }
 
-  toString () {
-    const mhStr = mh.toB58String(this.multihash)
-    return `DAGLink <${mhStr} - name: "${this.name}", size: ${this.size}>`
+  get name () {
+    return this._name
+  }
+
+  set name (name) {
+    throw ImmutableError
+  }
+
+  get size () {
+    return this._size
+  }
+
+  set size (size) {
+    throw ImmutableError
+  }
+
+  get multihash () {
+    return this._multihash
+  }
+
+  set multihash (multihash) {
+    throw ImmutableError
```

Comment:
@dignifiedquire not sure if it applies to errors in the same way, but yeah, your suggestion above -- https://github.com/ipld/js-ipld-dag-pb/pull/7#discussion_r88208831 -- makes sense.